src/detect-engine.c

```diff
@@ -2576,6 +2576,7 @@ DetectEngineCtx *DetectEngineCtxInitWithPrefix(const char *prefix, uint32_t tena
 static void DetectEngineCtxFreeThreadKeywordData(DetectEngineCtx *de_ctx)
 {
     HashListTableFree(de_ctx->keyword_hash);
+    de_ctx->keyword_hash = NULL;
```

Comment:
Why this change ? I mean what is the scenario that would lead to a use after free ?

Reply:
The "free" function for the `luaxform` transform is invoked after `GlobalsDestroy` is called. 

Stack when `GlobalDestroy` is invoked (and `keyword_hash` is cleared)
```
#0  DetectEngineCtxFreeThreadKeywordData (de_ctx=0x555557787020) at detect-engine.c:2578
#1  0x000055555583b45a in DetectEngineCtxFree (de_ctx=0x555557787020) at detect-engine.c:2652
#2  0x000055555583f43b in DetectEnginePruneFreeList () at detect-engine.c:4694
#3  0x000055555583f51a in DetectEngineClearMaster () at detect-engine.c:4719
#4  0x000055555569e8d4 in GlobalsDestroy () at suricata.c:384
#5  0x000055555569e48b in main (argc=18, argv=0x7fffffffdec8) at main.c:63
```

Stack when "free" is called for `luaxform` (and the `keyword_hash` is referenced again -- this time NULL with the newly added line
```
#0  0x0000555555ab332c in HashListTableRemove (ht=0x555556756840, data=0x7fffffffdb28, datalen=0) at util-hashlist.c:238
#1  0x000055555583d214 in DetectUnregisterThreadCtxFuncs (de_ctx=0x555557787020, data=0x555557816e30, name=0x55555610e8d1 "luaxform") at detect-engine.c:3671
#2  0x00005555559cbdc7 in DetectTransformLuaxformFree (de_ctx=0x555557787020, ptr=0x555557816e30) at detect-transform-luaxform.c:68
#3  0x0000555555841553 in DetectBufferTypeFreeFunc (ctx=0x555557787020, data=0x55555781a570) at detect-engine.c:976
#4  0x0000555555ab2fa8 in HashListTableFreeWithCtx (ctx=0x555557787020, ht=0x5555577c60b0) at util-hashlist.c:153
#5  0x000055555583b73a in DetectBufferTypeFreeDetectEngine (de_ctx=0x555557787020) at detect-engine.c:1764
#6  0x000055555583b4f6 in DetectEngineCtxFree (de_ctx=0x555557787020) at detect-engine.c:2674
#7  0x000055555583f43b in DetectEnginePruneFreeList () at detect-engine.c:4694
#8  0x000055555583f51a in DetectEngineClearMaster () at detect-engine.c:4719
#9  0x000055555569e8d4 in GlobalsDestroy () at suricata.c:384
#10 0x000055555569e48b in main (argc=18, argv=0x7fffffffdec8) at main.c:63
```

Reply:
Interesting...

Does it mean that `DetectEngineCtxFree` should call `DetectEngineCtxFreeThreadKeywordData` after `DetectBufferTypeFreeDetectEngine` ? (not sure this matters very much)

Reply:
Good suggestion; i'll try that.